### PR TITLE
Catlearn 15 match negatives

### PR DESCRIPTION
**Pull request by Kuraao**
Adding option to match orphaned relations of a cache negatively. I kept it as optional as I thought we might need to have the previous version of the match for some evaluations.
A class NegativeMatch is added to wrap labels of negative matches.

The following happens when matching with negatives:
- all arrows are negatively matched first. 
- Then labels of the test graph are tested against relations; the best relation is chosen according to minimization of: kl(relation, label) - kl(relation, negative)
- the relations which have been matched positively have their negative match removed.

This is rebased on your branch, so review after merging it.

### CODE REVIEW

**Comment by Previsou** [comment]
Simply rebased on master

File: catlearn/relation_cache.py [hunk 1]
```diff
@@ -231,6 +294,15 @@ def match(self, labels: DirectedGraph[NodeType]) -> Tsor:
...
         """
         result_graph = DirectedGraph[NodeType]()
+        if match_negatives:
+            for arr in self.arrows():
+                if not result_graph.has_edge(arr[0], arr[-1]):
+                    result_graph.add_edge(arr[0], arr[-1])
+                score = self[arr]
+                result_graph[arr[0]][arr[-1]][
+                    NegativeMatch(arr.derive())
+                ] = arr.derive(), kl_match(score)
```
Comment (Previsou) [inline catlearn/relation_cache.py:304]: I'd rather you specify explictly against_negative here. Less surprising imho

File: catlearn/relation_cache.py [hunk 2]
```diff
@@ -241,8 +313,18 @@ def match(self, labels: DirectedGraph[NodeType]) -> Tsor:
             if (
                     not self.graph.has_edge(src, tar)
```
Comment (Previsou) [inline catlearn/relation_cache.py:314]: aren't the two conditions equivalent?

File: catlearn/relation_cache.py [hunk 3]
```diff
@@ -231,6 +294,15 @@ def match(self, labels: DirectedGraph[NodeType]) -> Tsor:
         are added to the cache.
         """
         result_graph = DirectedGraph[NodeType]()
+        if match_negatives:
+            for arr in self.arrows():
+                if not result_graph.has_edge(arr[0], arr[-1]):
+                    result_graph.add_edge(arr[0], arr[-1])
+                score = self[arr]
```
Comment (Previsou) [inline catlearn/relation_cache.py:301]: Align with what you do below: score = kl_match(self[arr])

Comment (Previsou) [reply catlearn/relation_cache.py:301]: And avoid computing twice the derived arrow

File: catlearn/relation_cache.py [hunk 4]
```diff
@@ -241,8 +313,18 @@ def match(self, labels: DirectedGraph[NodeType]) -> Tsor:
...
                     not self.graph.has_edge(src, tar)
                     or not self.graph[src][tar]):
+
                 for arr in self.label_universe:
-                    self.add(CompositeArrow([src, tar], [arr]))
+                    # add arrow
+                    new_arr = CompositeArrow([src, tar], [arr])
+                    self.add(new_arr)
+
+                    # match new arrow against negative label
```
Comment (Previsou) [inline catlearn/relation_cache.py:322]: Add only if matching negatives?

File: tests/test_categorical_model.py [hunk 5]
```diff
@@ -306,17 +312,41 @@ def test_matching(
                         for arr in cache.label_universe}
 
                 kldivs = {
-                    idx: subproba_kl_div(score, value)
+                    idx: (
+                        subproba_kl_div(score, value)
```
Comment (Previsou) [inline tests/test_categorical_model.py:315]: Why not use kl_match directly?

File: tests/test_categorical_model.py [hunk 6]
```diff
@@ -306,17 +312,41 @@ def test_matching(
...
+                        - float(match_negatives) * subproba_kl_div(
+                            score, torch.zeros(score.shape)
+                            ))
                     for idx, score in available.items()}
 
                 expected_match = min(kldivs, key=kldivs.get)
-                expected_cost = kldivs[expected_match]
+                expected_score = available[expected_match]
+                expected_cost = subproba_kl_div(
```
Comment (Previsou) [inline tests/test_categorical_model.py:323]: How about using kl_match(expected_score, value, False)?

**Comment by Previsou** [comment]
Perhaps we should also test with match_negative fixture in TestDecisionCatModel & TestTrainableDecisionCatModel?

File: catlearn/relation_cache.py [hunk 7]
```diff
@@ -15,6 +16,66 @@
...
+    if against_negative is True, will return the log kl ratio of the score
+    for the given label against match for negative label
+    if against_negative is False, will return the kl of the score
+    for the given label
+
+    Additionally if label is not provided or is None,
+    will match against negative label
+    """
+    label_vector = torch.zeros(score.shape) if label is None else label
+    kl_div = subproba_kl_div(score, label_vector)
```
[the remark this answers is not in the document]
Comment (Previsou) [reply catlearn/relation_cache.py:33]: Answered my own question. The Pytorch doc could use some improvements here... First argument is expected to be log-proba, second one raw proba)

File: catlearn/relation_cache.py [hunk 8]
```diff
@@ -15,6 +16,66 @@
...
+    if against_negative:
+        return kl_div - kl_match(score)
```
[the remark this answers is not in the document]
Comment (Kuraao) [reply catlearn/relation_cache.py:35]: This is only used for ordering the scores. This is not the score which is actually written.

File: catlearn/relation_cache.py (hunk 4 above)
Comment (Kuraao) [reply catlearn/relation_cache.py:322]: Why?

File: tests/test_categorical_model.py (hunk 5 above)
Comment (Kuraao) [reply tests/test_categorical_model.py:315]: I didn't want to use it because it is defined in the relation_cache file (and this test class is supposed to be for this file). 
If I use it then I should add tests for the kl_match function.

File: tests/test_categorical_model.py (hunk 6 above)
Comment (Kuraao) [reply tests/test_categorical_model.py:323]: same as above

File: catlearn/relation_cache.py [hunk 9]
```diff
@@ -15,6 +16,66 @@
 BinaryOp = Callable[[Tsor, Tsor], Tsor]
 
 
+def kl_match(
+        score: Tsor, label: Optional[Tsor] = None,
+        against_negative: bool = False) -> Tsor:
+    """
+    Kullback-liebler divergence based match. Two modes can be used:
+    if against_negative is True, will return the log kl ratio of the score
+    for the given label against match for negative label
```
[the remark this answers is not in the document]
Comment (Kuraao) [reply catlearn/relation_cache.py:25]: Indeed, but kl_match is really just a convenience function for factoring some stuff though. I could rename it _kl_match to make it private ?

File: catlearn/relation_cache.py [hunk 10]
```diff
@@ -15,6 +16,66 @@
 BinaryOp = Callable[[Tsor, Tsor], Tsor]
 
 
+def kl_match(
+        score: Tsor, label: Optional[Tsor] = None,
+        against_negative: bool = False) -> Tsor:
```
[the remark this answers is not in the document]
Comment (Kuraao) [reply catlearn/relation_cache.py:21]: Why? bool is an immutable type, so handling the default case using None is weird

File: catlearn/relation_cache.py (hunk 4 above)
Comment (Kuraao) [reply catlearn/relation_cache.py:322]: Oh, what is after the comment..  Indeed, Sorry about that
(Was looking at what was before it so I didn't get it)

File: catlearn/relation_cache.py [hunk 11]
```diff
@@ -15,6 +16,67 @@
...
+    """
+    Kullback-Leibler divergence based match. Two modes can be used:
+    if against_negative is True, will return the the log of the ratio of the
```
Comment (Previsou) [inline catlearn/relation_cache.py:24]: This sentence still make no sense to me, at least not matching what is returned imho.
NB: also typo: 'the the'
